src/effects/effectslot.cpp

```diff
@@ -2,258 +2,528 @@
+        const unsigned int iEffectnumber,
+        EffectChain* pChainSlot,
+        EngineEffectChain* pEngineEffectChain)
+        : m_iEffectNumber(iEffectnumber),
+          m_group(group),
+          m_pEffectsManager(pEffectsManager),
+          m_pPresetManager(pEffectsManager->getEffectPresetManager()),
+          m_pBackendManager(pEffectsManager->getBackendManager()),
+          m_pMessenger(pEffectsMessenger),
+          m_pVisibleEffects(m_pEffectsManager->getVisibleEffectsList()),
+          m_pChain(pChainSlot),
+          m_pEngineEffectChain(pEngineEffectChain),
+          m_pEngineEffect(nullptr) {
+    VERIFY_OR_DEBUG_ASSERT(m_pEngineEffectChain) {
+        return;
+    }
 
-    m_pControlNumParameterSlots = new ControlObject(ConfigKey(m_group, "num_parameterslots"));
-    m_pControlNumParameterSlots->setReadOnly();
+    m_pControlLoaded = std::make_unique<ControlObject>(ConfigKey(m_group, "loaded"));
+    m_pControlLoaded->setReadOnly();
 
-    m_pControlNumButtonParameters = new ControlObject(ConfigKey(m_group, "num_button_parameters"));
-    m_pControlNumButtonParameters->setReadOnly();
+    m_pControlNumParameters.insert(EffectParameterType::KNOB,
+            QSharedPointer<ControlObject>(
+                    new ControlObject(ConfigKey(m_group, "num_parameters"))));
+    m_pControlNumParameters.insert(EffectParameterType::BUTTON,
+            QSharedPointer<ControlObject>(
+                    new ControlObject(ConfigKey(m_group, "num_button_parameters"))));
+    for (const auto& pControlNumParameters : std::as_const(m_pControlNumParameters)) {
+        pControlNumParameters->setReadOnly();
+    }
 
-    m_pControlNumButtonParameterSlots = new ControlObject(ConfigKey(m_group, "num_button_parameterslots"));
-    m_pControlNumButtonParameterSlots->setReadOnly();
+    m_pControlNumParameterSlots.insert(EffectParameterType::KNOB,
+            QSharedPointer<ControlObject>(
+                    new ControlObject(ConfigKey(m_group, "num_parameterslots"))));
+    m_pControlNumParameterSlots.insert(EffectParameterType::BUTTON,
+            QSharedPointer<ControlObject>(
+                    new ControlObject(ConfigKey(m_group, "num_button_parameterslots"))));
+    for (const auto& pControlNumParameterSlots : std::as_const(m_pControlNumParameterSlots)) {
+        pControlNumParameterSlots->setReadOnly();
+    }
 
     // Default to disabled to prevent accidental activation of effects
     // at the beginning of a set.
-    m_pControlEnabled = new ControlPushButton(ConfigKey(m_group, "enabled"));
+    m_pControlEnabled = std::make_unique<ControlPushButton>(ConfigKey(m_group, "enabled"));
     m_pControlEnabled->setButtonMode(ControlPushButton::POWERWINDOW);
-    connect(m_pControlEnabled, &ControlPushButton::valueChanged, this, &EffectSlot::slotEnabled);
+    connect(m_pControlEnabled.get(),
+            &ControlObject::valueChanged,
+            this,
+            &EffectSlot::updateEngineState);
```

Comment:
It is possible by using the valueChangeRequest signal, checking if all prerequisites are met and then using `setAndConfirm`.

Reply:
EffectSlots have always been able to be enabled without an effect loaded. I don't think that is a problem.

Reply:
Even if we do decide to change that, I don't think it should be done in this PR.

Reply:
I'm okay with not doing it in this PR, but I'd like to see it in a follow up. As you're short on time, I'd volunteer to fix those issues once this PR is merged.

Reply:
I don't agree that it is an issue. I can't recall ever experiencing a problem because of this nor anyone asking about it.

Reply:
Well, as I said, IMO its an invalid state (for an effect unit to be enabled, but no effect actually being present) and has some visual sideeffects, such as the unit seeming enabled in the UI or worse on a controller (since on a controller you don't even notice that there is no effect loaded, you just see the slot is enabled). Its a relatively straightforward fix, but if you don't agree its a bug worth fixing, I'd take this discussion to zulip to get the others opinion on it.

Reply:
I think we should fix it unless there is some legitimate use case for it.

One thing I could think off:
If you use SHIFT+FX Knob to select an effect while it's active, will such a fix lead to the FX being disabled when it reaches the "No effect" item? That would make it necessary to enable it again, which would be clunky if you're just trying different effects in your headphones.

Reply:
Yeah, we'd have to make an exception in this case...

Reply:
I don't know how such an exception should look without making it much more complex. But such a change should be discussed separately and not rushed into this PR.

Reply:
Yes, let's continue this discussion on Zulip.

Reply:
I agree, as I said its not a blocker for this PR and I'm offering to look into solutions once this PR is merged.